### PR TITLE
Use a CommaDelimitedList for Environment variables and Secrets

### PR DESCRIPTION
**Pull request by ptmclean**
Hi, I'm hoping this is an improvement for allowing consumers of this module to pass a larger number of env vars to their fargate services without effecting the number of parameters used.

If you like this method I think it could be used to allow larger numbers of connections to be passed into modules, for example Dependencies.

In anycase let me know what you think.

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

**Comment by michaelwittig** [comment]
Hi @ptmclean We used a similar approach at the very beginning. Problems we were faced with in another project https://github.com/widdix/aws-cf-templates:
1. values with a `,` are not possible
2. CodePipeline's way to pass variable parameters into stacks is not able to use any string concatenation functions (this might not be a problem for cfn-modules because the modules are not created by codepipeline, only the parent stack)

**Comment by ptmclean** [comment]
Hmmm, Thanks for that. I suspected this approach would have been considered before. It sounds like these problems wouldn't be a problem for lists where the values are ARNs or StackNames like secrets or dependencies. So possibly could be considered for these?

For clear text env vars,  the limitation around commas is significant. From what I can see, commas used to be escapable when passed in the `aws cloudformation create-stack` but I can't find a way using the `--override-parameters` syntax of `deploy`.

We could make change the delimiters to something less used. It would be less intuitive but the docs could cover the that. Something like...
```
AppEnvironmentVariables: 'Key1==Value with a ,|Key2==The second value with an ='
```

What do you think? We could just limit this change to secrets. A similar change could be quite valuable for the dependency pattern in the lambda-function module.

**Comment by michaelwittig** [comment]
What about an additional way using the proposed syntax? This way we avoid a breaking change, the old way is compatible with code deploy and works with all chars but we still have the easier to-use variant you suggested?

**Comment by ptmclean** [comment]
Ok, that sounds like a good idea. I've updated so please let me know what you think.

File: module.yml [hunk 1]
```diff
@@ -103,6 +103,14 @@ Parameters:
     Default: 80
     MinValue: 1
     MaxValue: 49150
+  AppEnvironmentVariables:
+    Description: 'An optional string representation of a list of environment variables. Each key value pair is separated by ,, and each key and value by ==. eg. "key1==value1,,key2==value2"'
+    Type: String
+    Default: 'END_MARKER'
```
Comment (michaelwittig) [inline module.yml:109]: wouldn't the default `''` work as well?

File: module.yml [hunk 2]
```diff
@@ -103,6 +103,14 @@ Parameters:
...
+  AppSecretVariables:
+    Description: 'An optional CommaDelimitedList of up to 10 Environment Variables with the value the name of a secret module. Eg. Key1=SecretModuleName'
+    Type: CommaDelimitedList
+    Default: 'END_MARKER'
```
Comment (michaelwittig) [inline module.yml:113]: wouldn't the default `''` work as well?

File: module.yml [hunk 3]
```diff
@@ -103,6 +103,14 @@ Parameters:
     Default: 80
     MinValue: 1
     MaxValue: 49150
+  AppEnvironmentVariables:
+    Description: 'An optional string representation of a list of environment variables. Each key value pair is separated by ,, and each key and value by ==. eg. "key1==value1,,key2==value2"'
```
Comment (michaelwittig) [inline module.yml:107]: I'm thinking if we should keep the syntax the same as in AppSecretVariables: `key=value` ? If the user has special chars, the "old" way ist still available?

**Review by michaelwittig — CHANGES_REQUESTED** [review]

**Comment by michaelwittig** [comment]
test

File: module.yml (hunk 3 above)
Comment (ptmclean) [reply module.yml:107]: Sorry I've been slow on these. I can take a look at this tonight, I'm wondering about this one though. You're right that it is better not to have breaking changes, but wouldn't it be better for consumers not to have to use 2 different ways to add environment variables?

File: module.yml (hunk 2 above)
Comment (ptmclean) [reply module.yml:113]: I'll check that out. I certainly needed this Default value for some reason but there have been a few changes so perhaps not.

File: module.yml (hunk 3 above)
Comment (ambsw-technology) [reply module.yml:107]: For backwards compatibility you normally:

 - Introduce the new format
 - Deprecate the old one (in the docs and template)
 - After some time fully remove the old approach from the template

**Comment by michaelwittig** [comment]
I wonder if we could use the new `environmentFiles` feature https://docs.aws.amazon.com/AmazonECS/latest/developerguide/taskdef-envfiles.html

**Pull request closed without merging**